omegaconf/omegaconf.py

```diff
@@ -124,17 +126,13 @@ def env(key: str, default: Any = _EMPTY_MARKER_, *, config: BaseContainer) -> An
                 return val_str
 
         # Resolve the parse tree.
-        val = config.resolve_parse_tree(parse_tree)
+        assert parent is not None
+        root = OmegaConf.get_root(parent)
+        val = root.resolve_parse_tree(parse_tree)
         return _get_value(val)
 
     # Note that the `env` resolver does *NOT* use the cache.
-    OmegaConf.register_resolver(
-        "env",
-        env,
-        config_arg="config",
-        args_as_strings=False,
-        use_cache=False,
-    )
+    OmegaConf.new_register_resolver("env", env, use_cache=False)
```

Comment:
Why is the env resolver not caching? is this in line with the behavior in 2.0?

Reply:
> Why is the env resolver not caching? is this in line with the behavior in 2.0?

No this is a breaking change (listed in a news item, once you get to it).

Originally the motivation came from my previous implementation where I was systematically disabling the cache for any resolver that would directly access the config. However, as argued in https://github.com/omry/omegaconf/pull/321#discussion_r472104910, I also think it makes more sense (see the example in that comment) so I intended to keep that change (in addition, I see some value in allowing someone to control a config option via an env variable).
I could revert it though if you prefer.

Reply:
I think we should not change the behavior of the env resolver. we can potentially introduce a second one (uncached_env?) if there is demand.

Reply:
> I think we should not change the behavior of the env resolver. we can potentially introduce a second one (uncached_env?) if there is demand.

Ok, I don't have a strong opinion about this, I don't think we need to register a second one (a user can easily do that themselves if they need it). I'll just wait until we decide the path to go in https://github.com/omry/omegaconf/pull/321#discussion_r527345832 before changing it to use the cache again.

Reply:
Okay, let's keep it cached then.

Reply:
Done in fa07d48